titan-compiler/coder.lua

```diff
@@ -645,127 +733,275 @@ local function rec_declare_struct(rec)
             $FIELDS
         };
     ]], {
-        NAME = rec_struct_name(rec),
+        NAME = self:struct(),
         FIELDS = table.concat(fields, "\n"),
     })
     return out
 end
 
-local function rec_gc_slot(rec, udata, field_name)
+function RecordCoder:gc_slot(udata, field_name)
     local out = util.render([[&${UDATA}->uv[${I}].uv]], {
         UDATA = udata,
-        I = rec._layout.gc_pos[field_name] - 1,
+        I = self.layout.gc_pos[field_name] - 1,
     })
     return out
 end
 
-local function rec_primitive_slot(rec, udata, field_name)
+function RecordCoder:primitive_slot(udata, field_name)
     local out = util.render(
         [[((${STRUCT_NAME} *)getudatamem(${UDATA}))->${FIELD_NAME}]],
     {
-        STRUCT_NAME = rec_struct_name(rec),
+        STRUCT_NAME = self:struct(),
         UDATA = udata,
-        FIELD_NAME = rec_field_name(rec, field_name),
+        FIELD_NAME = self:field(field_name),
     })
     return out
 end
 
-local function rec_set_field(rec, udata, field_name, cvalue)
-    local typ = rec._field_types[field_name]
+function RecordCoder:set_field(udata, field_name, cvalue)
+    local typ = self.tl_node._field_types[field_name]
     if types.is_gc(typ) then
-        local slot = rec_gc_slot(rec, udata, field_name)
+        local slot = self:gc_slot(udata, field_name)
         return set_heap_slot(typ, slot, cvalue, udata)
     else
         local out = util.render([[
             ${SLOT} = ${CVALUE};
         ]], {
-            SLOT = rec_primitive_slot(rec, udata, field_name),
+            SLOT = self:primitive_slot(udata, field_name),
             CVALUE = cvalue,
         })
         return out
     end
 end
 
-local function rec_metatable_type()
-    -- TODO: this should be types.T.Table()
+function RecordCoder:dyn_dispatch_linear(fields, key, action)
+    local cstats = {}
+
+    for i, field in ipairs(fields) do
+        table.insert(cstats, util.render([[
+            ${IF} (strcmp(${KEY}, ${FIELD}) == 0) {
+                ${ACTION}
+            }
+        ]], {
+            IF = i == 1 and "if" or "else if",
+            KEY = key,
+            FIELD = c_string(field.name),
+            ACTION = action(field.name),
+        }))
+    end
+
+    table.insert(cstats, util.render([[
+        ${ELSE} {
+            titan_runtime_record_index_error(L, ${KEY});
+        }
+    ]], {
+        ELSE = #fields > 0 and "else" or "",
+        KEY = key,
+    }))
+
+    return table.concat(cstats, "\n")
+end
+
+function RecordCoder:dyn_dispatch_bin(fields, key, action, ctx)
```

Comment:
What about receiving `fields`, `i` and `j` as a parameter instead of using a slice function to create lots of copies of the list?